### ISSUE TITLE
this.disposable is undefined

### ISSUE THREAD

**Issue by jasonkuhrt** [cujojs/most#426]
### Summary

I have run into a case where the typeof check below returns `undefined`:

```js
SliceSink.prototype.dispose = function () {
  console.log("calling dispose", typeof this.disposable)
  console.log("this", this)
  return this.disposable.dispose()
}
```
```
TypeError: Cannot read property 'dispose' of undefined
    at SliceSink.dispose (build/main.js:20834:25)
    at SliceSink.event (build/main.js:20826:10)
    at tryEvent (build/main.js:2483:10)
    at SubscriberSink.next (build/main.js:22463:92)
    at SubscriptionObserver.next (build/main.js:23958:14)
    at observerNext (build/main.js:6253:16)
    at addListener (build/main.js:18989:17)
    at attachListenerToElement (build/main.js:19101:13)
    at utils.forEach (build/main.js:5604:22)
    at Object.listenTo (build/main.js:19025:9)
```

### Expected result

this.disposable should never be undefined or a better error for the user.
### Actual Result

this.disposable is undefined

### Versions

- most.js: ^1.2.2

### Steps to reproduce

See code below

### Code to reproduce

```js
const FRP = require("most")
const Observable = require("zen-observable")



const makeProblemStream = () =>
  new Observable((observer) => {
    observer.next(1) // <-- Triggers error
    const intervalId = setInterval(
      () => { observer.next(1) },
      1000
    )
    return () => clearInterval(intervalId)
  })

FRP
.from(makeProblemStream())
.take(1)
.drain()
.catch(console.error)

```

**Comment by jasonkuhrt** [comment]
If I update my code to this it works, but I have no idea why, nor does it invalidate the reported issue. But interesting none the less and worth keeping in mind:

```js
  it("can re-observe", () => {
    let el = makeDiv()
    temporary.appendChild(el)
    const stream = Dom.observeDomEvent("resize", el)
    return FRP
      .from(stream)
      .take(1)
      .drain()
      .then(() => {
        el = makeDiv()
        temporary.appendChild(el)
        const stream2 = Dom.observeDomEvent("resize", el)
        return FRP
          .from(stream2)
          .take(1) // <-- Error here
          .drain()
      })
  })
```

**Comment by jasonkuhrt** [comment]
I found another way to avoid the error that is more telling:

```js
// dom.js
import ERD from "element-resize-detector"
import Observable from "zen-observable"

const observeDomEvent = (eventName, element) => {
  return new Observable((observer) => {
    const erd = ERD({ strategy: "scroll" })
    const observerNext = (event) => {
      setTimeout(() => {
        observer.next(event)
      }, 0)
    }
    const cleanUp = () => {
      erd.removeListener(element, observerNext)
    }
    erd.listenTo(element, observerNext)
    return cleanUp
  })
}
```

**Comment by jasonkuhrt** [comment]
Continuing to reduce the repro, here's a minimal version now:

```js
const observeDomEvent = () =>
  new Observable((observer) => {
    observer.next(1)
    const intervalId = setInterval(
      () => { observer.next(1) },
      1000
    )
    return () => clearInterval(intervalId)
  })
```

**Comment by briancavalier** [comment]
Hey @jasonkuhrt.  I really appreciate your digging into this and trying to find a minimal case.  I'm not sure I quite understand your most recent comment, though: it doesn't seem to show any most.js code, so I'm trying to understand how it fits into the previous examples.  Do you have a complete/runnable example (maybe via webpackbin or esnextbin) for the latest one?  Thanks!

**Comment by jasonkuhrt** [comment]
@briancavalier Sorry about the confusion. Let me know if this is clearer:

```js
const FRP = require("most")
const Observable = require("zen-observable")



const makeProblemStream = () =>
  new Observable((observer) => {
    observer.next(1) // <-- Triggers error
    const intervalId = setInterval(
      () => { observer.next(1) },
      1000
    )
    return () => clearInterval(intervalId)
  })

FRP
.from(makeProblemStream())
.take(1)
.drain()
.catch(console.error)
```
```
❯ node repro.js
TypeError: Cannot read property 'dispose' of undefined
    at SliceSink.dispose (/Users/jason.kuhrt/projects/forto/node_modules/most/lib/combinator/slice.js:123:25)
    at SliceSink.event (/Users/jason.kuhrt/projects/forto/node_modules/most/lib/combinator/slice.js:117:10)
    at tryEvent (/Users/jason.kuhrt/projects/forto/node_modules/most/lib/source/tryEvent.js:14:10)
    at SubscriberSink.next (/Users/jason.kuhrt/projects/forto/node_modules/most/lib/observable/fromObservable.js:51:26)
    at Object.next (/Users/jason.kuhrt/projects/forto/node_modules/zen-observable/zen-observable.js:154:14)
    at Observable (/Users/jason.kuhrt/projects/forto/repro.js:8:14)
    at Object.Subscription (/Users/jason.kuhrt/projects/forto/node_modules/zen-observable/zen-observable.js:103:32)
    at Observable.subscribe (/Users/jason.kuhrt/projects/forto/node_modules/zen-observable/zen-observable.js:229:12)
    at ObservableSource.run (/Users/jason.kuhrt/projects/forto/node_modules/most/lib/observable/fromObservable.js:35:29)
    at new SliceSink (/Users/jason.kuhrt/projects/forto/node_modules/most/lib/combinator/slice.js:97:41)
```

**Comment by briancavalier** [comment]
Great, thanks @jasonkuhrt.  I'll start to dig into this.

**Comment by briancavalier** [comment]
@jasonkuhrt There appears to be a bug in take and takeWhile that causes it to dispose the underlying stream twice.  There's a candidate fix in [this branch](https://github.com/cujojs/most/tree/fix-take-double-dispose) that works for the simple repro.js above (repro.js was super-helpful, thank you again!), and still passes all existing unit tests.  Could you try it in your larger use case to see if it helps there as well?

If so, then I'll try to get this released after seeing if the unit tests could have been improved to catch it.

Thanks!

**Comment by jasonkuhrt** [comment]
> Could you try it in your larger use case to see if it helps there as well?

Hey @briancavalier I just tried out your branch in [forto](https://github.com/jasonkuhrt/forto) and it works!

On a side note, just wanted say thanks for your fantastic OSS contributions. More over, you conduct yourself with an amazing humility and humanity... its always a pleasure to interact with you!

**Comment by brandonpayton** [comment]
Thanks for reporting this @jasonkuhrt. I ran into the same thing on a pet project but didn't have time to look into it.

Thank you for the quick response @briancavalier.

**Comment by briancavalier** [comment]
Thank you for those very kind words, @jasonkuhrt!

See #428 for the cleaned-up fix (yay fix by deleting code 😄 ).  Once that's merged, we can release 1.2.3.
